### ISSUE TITLE
[Safe Search] Javascript script to hide results

### ISSUE THREAD

**Issue by GreenLunar** [searx/searx#636]
Safe search does not always work.  I suggest to provide a Javascript script, for client-side, that would hide suspected results.

**Comment by pointhi** [comment]
?

safe search is implemented on the engines. When the engine doesn't support safesearch or the engine has other ideas how safe search should work than you it can happen to show results which aren't "save".

Even then, what can JavaScript do which python can't do?

**Comment by GreenLunar** [comment]
Try searching for Doggy or Hamster.

JavaScript script, even though can be disabled, can run on client-side, which consequently saves load on server-side.

**Comment by kvch** [comment]
@GreenLunar I don't understand your reply. Could you elaborate?

**Comment by GreenLunar** [comment]
Even though users can disable JavaScript, JavaScript script can run on client-side; Python, however, can only run on server-side.

If you want to do it in Python, I have no problem with that.

**Comment by pointhi** [comment]
that comment doesn't make sense at all:
- a JS implementation wouldn't be able to be used by all users
- a JS implementation increases bandwith requirements of the server, especially problematic for low bandwith users
- a JS implementation would remove results not until the latest state of processing, instead of deleting them right after retrieving the results from the engine when done in python.
- how would you even detect suspect results using JS? Downloading a blocklist every time?
- that feature doesn't need client side code at all

**Comment by kvch** [comment]
I totally agree with @pointhi. Safesearch works right like it should. I think this issue is invalid.

**Comment by GreenLunar** [comment]
So be it in Python on server-side.  @kvch, why do you think this issue is invalid?

> how would you even detect suspect results using JS? Downloading a blocklist every time?

Yes, a blocklist in a form of a keywords dictionary, which may be stored in offline cache.

**Comment by kvch** [comment]
Because you have written that 

> Safe search does not always work.

As I have written before, it works perfectly. It is not searx's fault that a few search engines "safe results" are not safe enough. Searx only proxies the results from other services which implement safe search.

Furthermore, as @pointhi pointed out

> a JS implementation wouldn't be able to be used by all users

And also the performance and the sanity your initial proposal was not sufficient.

**Comment by dalf** [comment]
I haven't checked in details, but I think that in the current implementation if the safe search setting is "moderate" or "strict", the search is still done through all engines even engines that doesn't support safe search.

So, that's mean that if the safe search is not implemented for an engine and that engine can provide unsafe contents, the "safe search" doesn't work from the user point of view. 

See : https://github.com/asciimoo/searx/blob/master/searx/search.py#L295 
there is no "skip the engine if it doesn't support safe search and if the user ask for a safe search"

A way to improve that : 
- skip the engine that doesn't support safe-search as it is done in the time range PR : https://github.com/asciimoo/searx/pull/634/files#diff-35de57dede17c7660869ae806bbc99b6R283
- Some search engines will not provide unsafe content for sure (photon, osm, wolframalpha engine are in this category). The implementation can say that "safe search" is supported, or make a distinction with a special value (False, True, "SAFE" or "NA" for example).

**Comment by GreenLunar** [comment]
Shall we close this issue?

**Comment by Quix0r** [comment]
"Safe-search" is some kind of "self-censorship", means not wanting to see, e.g. NSFW content (e.g. on workspace). It is okay for me to have it included in the engine (Python-side) but I guess it is an endless battle ("cat-mouse battle") as marketeers get trickier into slipping in bogus/low-quality/adult-only results to target searx or any other "non-wished" search engine.

Still I would find it okay, to let the user decide to have NSFW results (not just erotic/pr0n) in. :-)

**Comment by Quix0r** [comment]
So, Python-side is enforced on the end-user/admin while the admin might be able to still control it (can be risky in some restrictive countries, but his/her choice) and it might be exposed to the end-user in the `about` page.

**Comment by Quix0r** [comment]
To clarify myself, I'm also **against** censorship. :-)

**Comment by GreenLunar** [comment]
> Still I would find it okay, to let the user decide to have NSFW results (not just erotic/pr0n) in. :-)

Ditto!

**Comment by sjehuda** [comment]
> * a JS implementation would remove results not until the latest state of processing, instead of deleting them right after retrieving the results from the engine when done in python.

Yes, this is true.
You can use JS as a complement to server-side safe search.

> * how would you even detect suspect results using JS? Downloading a blocklist every time?

Yes, and caching that blacklist.
I'm assuming the blacklists are devided by locale  and each blacklist is concise and namely based on keywords which will make it shorter per blacklist.
I don't think it'd produce a load higher than the counterparts of Searx.
